File: net/Azure.Identity.Extensions/StaticTokenCredential.cs
using Azure.Core;
using Azure.Identity.Extensions.implementation;
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace Azure.Identity.Extensions
{
    public class StaticTokenCredential : TokenCredential
    {
        private readonly AccessToken _accessToken;
        private readonly CredentialPipeline _pipeline;

        /// <summary>
        /// Protected constructor for mocking.
        /// </summary>
        protected StaticTokenCredential()
        {
        }

        /// <summary>
        /// Creates an instance of the StaticTokenCredential.
        /// </summary>
        /// <param name="tokenString">tokenString The string of prefetched token.</param>
        public StaticTokenCredential(string tokenString)
            : this(tokenString, null)
        {
        }

        /// <summary>
        /// Creates an instance of the StaticTokenCredential.
        /// </summary>
        /// <param name="accessToken">accessToken The prefetched token.</param>
        public StaticTokenCredential(AccessToken accessToken)
           : this(accessToken, null)
        {
        }

        /// <summary>
        /// Creates an instance of the StaticTokenCredential.
        /// </summary>
        /// <param name="tokenString">tokenString The string of prefetched token.</param>
        /// <param name="options">Options that allow to configure the management of the requests sent to the Azure Active Directory service.</param>
        public StaticTokenCredential(string tokenString, TokenCredentialOptions options)
            : this(tokenString, default, options, null)
        {
        }

        /// <summary>
        /// Creates an instance of the StaticTokenCredential.
        /// </summary>
        /// <param name="accessToken">accessToken The prefetched token.</param>
        /// <param name="options">Options that allow to configure the management of the requests sent to the Azure Active Directory service.</param>
        public StaticTokenCredential(AccessToken accessToken, TokenCredentialOptions options)
            : this (null, accessToken, options, null)
        {
        }

        internal StaticTokenCredential(string tokenString, AccessToken accessToken, TokenCredentialOptions options, CredentialPipeline pipeline)
        {
            _accessToken = new AccessToken(
            tokenString != null ? tokenString : accessToken.Token,
            tokenString != null ? DateTimeOffset.Now.AddDays(1).UtcDateTime : accessToken.ExpiresOn
        );
            _pipeline = pipeline ?? CredentialPipeline.GetInstance(options);
        }

        /// <summary>
        /// Obtains a token from the Azure Active Directory service, using the specified client secret to authenticate. This method is called automatically by Azure SDK client libraries. You may call this method directly, but you must also handle token caching and token refreshing.
        /// </summary>
        /// <param name="requestContext">The details of the authentication request.</param>
        /// <param name="cancellationToken">A <see cref="CancellationToken"/> controlling the request lifetime.</param>
        /// <returns>An <see cref="AccessToken"/> which can be used to authenticate service client calls.</returns>
        public override async ValueTask<AccessToken> GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken = default)
        {
            using CredentialDiagnosticScope scope = _pipeline.StartGetTokenScope("StaticTokenCredential.GetToken", requestContext);

            try
            {
                return await Task.FromResult(scope.Succeeded(_accessToken));
            }
            catch (Exception e)
            {
                throw scope.FailWrapAndThrow(e);
            }
        }

        /// <summary>
        /// Obtains a token from the Azure Active Directory service, using the specified client secret to authenticate. This method is called automatically by Azure SDK client libraries. You may call this method directly, but you must also handle token caching and token refreshing.
        /// </summary>
        /// <param name="requestContext">The details of the authentication request.</param>
        /// <param name="cancellationToken">A <see cref="CancellationToken"/> controlling the request lifetime.</param>
        /// <returns>An <see cref="AccessToken"/> which can be used to authenticate service client calls.</returns>
        public override AccessToken GetToken(TokenRequestContext requestContext, CancellationToken cancellationToken = default)
        {
            using CredentialDiagnosticScope scope = _pipeline.StartGetTokenScope("StaticTokenCredential.GetToken", requestContext);

            try
            {
                return scope.Succeeded(_accessToken);
            }
            catch (Exception e)
            {
                throw scope.FailWrapAndThrow(e);
            }
        }
    }
}

File: go/azidext/static_token_credential.go
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package azidext

import (
	"context"
	"time"

	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
)

// StaticTokenCredential enables authentication to Azure Active Directory using a access token.
type StaticTokenCredential struct {
	accessToken *azcore.AccessToken
}

// NewStaticTokenCredential constructs a new StaticTokenCredential with the details needed to authenticate against Azure Active Directory with a access token.
// tenantID: The Azure Active Directory tenant (directory) ID of the service principal.
// clientID: The client (application) ID of the service principal.
// clientSecret: A client secret that was generated for the App Registration used to authenticate the client.
// options: allow to configure the management of the requests sent to Azure Active Directory.
func NewStaticTokenCredential(tokenString string, accessToken *azcore.AccessToken) *StaticTokenCredential {
	if tokenString != "" {
		converted := &azcore.AccessToken{
			Token:     tokenString,
			ExpiresOn: time.Now().AddDate(0, 0, 1).UTC(),
		}
		return &StaticTokenCredential{accessToken: converted}
	}

	return &StaticTokenCredential{accessToken: accessToken}
}

// GetToken obtains a token from Azure Active Directory, using the specified client secret to authenticate.
// ctx: Context used to control the request lifetime.
// opts: TokenRequestOptions contains the list of scopes for which the token will have access.
// Returns an AccessToken which can be used to authenticate service client calls.
func (c *StaticTokenCredential) GetToken(ctx context.Context, opts azcore.TokenRequestOptions) (*azcore.AccessToken, error) {
	logGetTokenSuccess(c, opts)
	return c.accessToken, nil
}

func (c *StaticTokenCredential) AuthenticationPolicy(options azcore.AuthenticationPolicyOptions) azcore.Policy {
	return newBearerTokenPolicy(c, options)
}
